enhancements/installer/signal-cluster-deletion.md

```diff
@@ -0,0 +1,152 @@
+
+## Motivation
+
+### Goals
+
+Signal that cluster deletion has begun through deletion of a new CRD `Alive` and wait for that object to be deleted during uninstall before continuing with removing cluster resources such as instances, storage and networking.
+
+### Non-Goals
+
+* Attaching finalizers to operator resources based on the `Alive` object which would facilitate removal of operator resources.
+* Removing resources other than the `Alive` object during cluster uninstall.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+As an OpenShift administrator, I would like for any external resources created in service of the cluster be removed when when I delete the cluster, so that:
+* I do not have to track down those resources and remove them manually
+* I do not have to pay for resources that are not being used.
+
+#### Story 2
+As an OpenShift (software, big-O) Operator, I would like to have a way to know that the cluster is being shut down, so that I can perform any necessary cleanup steps and remove external resources.
+
+#### Story 3
+As an OpenShift administrator, I would like to have a way to indicate that certain external resources that would normally be cleaned up on cluster deletion be preserved on cluster delete, so that I don't lose important resources stored externally.
+
+### Implementation Details/Notes/Constraints
+
+#### API
+
+Introduce a new CRD `Alive` from which operator resources could attach a finalizer initiating cascading delete when the `Alive` resource is deleted.
+
+```yaml
+apiVersion: v1alpha1
+kind: Alive
+metadata:
+  name: cluster
+  namespace: openshift-config
+spec:
+  blockTeardown: true  # block teardown of operator resources
+```
+
+Add an admission plugin that prevents delete when `blockTeardown` has been set.
+
+*Note*: We proposed adding the CRD and admission plugin (which pevents delete) to the cluster version operator and it was decided that they would be better placed within OLM. We would prefer not to create a separate operator for a CRD and admission plugin as that would require a repository and release engineering. We're requesting feedback on where to put these resources.
+
+#### Install
+
+Create a `Alive` object during installation.
+
+#### Upgrade
+
+Create `Alive` object for existing clusters.
+
+#### Uninstall
+
+Delete a cluster's `Alive` resource during uninstall when requested via flag such as `openshift-install destroy cluster --cluster-alive-delete`. Cluster destroy will wait for a default amount of time and fail if `Alive` deletion was not successful. The default timeout will not be configurable and users will be expected to attempt shutdown multiple times upon failure.
```

Comment:
Do enough OCP users use `openshift-install destroy` rather than deleting the relevant cluster resources directly to make this viable?  What about customers or automated systems (e.g. customers' CI systems) that might delete cluster resources directly rather than calling `openshift-install destroy`?

These arguments make me think that attempting to solve this problem inside the cluster might not be wise.

Reply:
FWIW, ARO does not use `openshift-install destroy` and would need implementation work to support this enhancement if it happened.

Reply:
I will rework this to be less openshift-install specific. We do leverage openshift-install destroy code in hive and indeed other provisioning infra would need to do the same, *if* they want to make use of this cleanup functionality.

It would be difficult to manage externally because we simply don't know the details for cleanup of random things that might be in the cluster. The idea here was to allow operators (or any other componen) to maintain it's own cleanup logic, we just want to signal to tell it a teardown is imminent.